api/client/client.go

```diff
@@ -326,22 +327,48 @@ type (
 	}
 )
 
-// authConnect connects to the Teleport Auth Server directly.
+// authConnect connects to the Teleport Auth Server directly or through Proxy.
 func authConnect(ctx context.Context, params connectParams) (*Client, error) {
-	dialer := NewDialer(ctx, params.cfg.KeepAlivePeriod, params.cfg.DialTimeout, WithTLSConfig(params.tlsConfig))
+	dialer := NewDialer(ctx, params.cfg.KeepAlivePeriod, params.cfg.DialTimeout,
+		WithInsecureSkipVerify(params.cfg.InsecureAddressDiscovery),
+		WithALPNConnUpgrade(IsWebProxyAndConnUpgradeRequired(ctx, params.addr, &params.cfg)),
+	)
+
 	clt := newClient(params.cfg, dialer, params.tlsConfig)
 	if err := clt.dialGRPC(ctx, params.addr); err != nil {
 		return nil, trace.Wrap(err, "failed to connect to addr %v as an auth server", params.addr)
 	}
 	return clt, nil
 }
 
+// IsWebProxyAndConnUpgradeRequired returns true if targetAddr is a Teleport
+// web proxy address and ALPN connection upgrade is required for it. If no
+// cluster name is provided for ALPN, assume dialer is not trying to connect
+// Auth through Proxy using TLS Routing.
+func IsWebProxyAndConnUpgradeRequired(ctx context.Context, targetAddr string, cfg *Config) bool {
+	return cfg.ALPNSNIAuthDialClusterName != "" &&
+		isWebProxy(ctx, targetAddr, cfg) &&
+		cfg.IsALPNConnUpgradeRequired(targetAddr, cfg.InsecureAddressDiscovery)
+}
+
+func isWebProxy(ctx context.Context, targetAddr string, cfg *Config) bool {
+	if cfg.WebProxyAddr != "" {
+		return cfg.WebProxyAddr == targetAddr
+	}
+	_, err := webclient.Find(&webclient.Config{
+		Context:   ctx,
+		ProxyAddr: targetAddr,
+		Insecure:  cfg.InsecureAddressDiscovery,
+	})
+	return err == nil
```

Comment:
We trust `cfg.WebProxyAddr` if it's provided. In case it's empty, we can try to web ping to find out. 

Or, I can remove this but it will require the caller to provide a `WebProxyAddr` when the cluster name for ALPN is also provided. What do you think?